### ISSUE TITLE
Boolean Operations Between Variables and PartialLink

### ISSUE THREAD

**Issue by log0ymxm** [AI-DI/Brancher#6]
Seems like lt, le, gt, ge, eq, ne are all useful operations in probabilistic models. I tried out adding these to the Variable & PartialLink classes but maybe there's more to it than that. Are these operations that are doable?

**Comment by LucaAmbrogioni** [comment]
I agree we should definitely implement them. They are definitely doable but there are subtleties that we need to take into account. 
It is easy to define logical operations between variables and PartialLinks that return Boolean values but  the trickier part is to use them for gating the computation graph. I think we need a Gate class as Boolean-valued subclass of PartialLink which in turn can gate other variables by multiplication.

For example, we could have expressions of this form:

a = RandomVariable, b = RandomVariable, c = RandomVariable, d = RandomVariable

e = (a > b)*c + (a<=b)*d

where (a>b) and (a<=b) are gates.

Of course we could reach the same goal by having (a>b) be a DeterministicNode with {0,1} values but this would be computationally inefficient as it will run the forward pass of the gated branch.

**Comment by log0ymxm** [comment]
Yeah, this would be my exact use case. Do other probabilistic frameworks optimize this kind of gating?
